File: src/main/java/io/github/hiiragi283/htms/api/HTMaterialsPlugin.java
package io.github.hiiragi283.htms.api;

import net.minecraft.util.ResourceLocation;

import org.jetbrains.annotations.NotNull;

import io.github.hiiragi283.htms.api.material.HTMaterialRegistry;
import io.github.hiiragi283.htms.api.material.content.HTMaterialContentRegistry;
import io.github.hiiragi283.htms.api.shape.HTShapeRegistry;

public interface HTMaterialsPlugin {

    @NotNull
    String getModId();

    default @NotNull ResourceLocation getId(@NotNull String path) {
        return new ResourceLocation(getModId(), path);
    }

    int getPriority();

    // Pre Init //

    default void registerShape(@NotNull HTShapeRegistry.Builder builder) {}

    default void registerMaterial(@NotNull HTMaterialRegistry.Builder builder) {}

    default void registerMaterialContent(@NotNull HTMaterialContentRegistry.Builder builder) {}
}

File: src/main/kotlin/io/github/hiiragi283/api/HTMaterialsPlugin.kt
package io.github.hiiragi283.api

import io.github.hiiragi283.api.material.HTMaterialRegistry
import io.github.hiiragi283.api.shape.HTShapeRegistry
import net.minecraftforge.fml.common.eventhandler.Event

interface HTMaterialsPlugin {

    val modId: String
    val priority: Int

    fun registerShape(builder: HTShapeRegistry.Builder) {}

    fun registerMaterial(builder: HTMaterialRegistry.Builder) {}

    fun afterMaterialRegistration(instance: HTMaterialsAPI, isClient: Boolean) {}

    class RegisterEvent(private val list: MutableList<HTMaterialsPlugin>) : Event() {

        fun register(plugin: HTMaterialsPlugin) {
            list.add(plugin)
        }
    }

}